File: weblayer/browser/java/org/chromium/weblayer_private/GoogleAccountsCallbackProxy.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.weblayer_private;

import android.os.RemoteException;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.components.signin.GAIAServiceType;
import org.chromium.weblayer_private.interfaces.GoogleAccountServiceType;
import org.chromium.weblayer_private.interfaces.IGoogleAccountsCallbackClient;

/**
 * Owns the C++ GoogleAccountsCallbackProxy which is responsible for forwarding all calls to this
 * class.
 */
@JNINamespace("weblayer")
public final class GoogleAccountsCallbackProxy {
    private long mNativeGoogleAccountsCallbackProxy;
    private IGoogleAccountsCallbackClient mClient;

    GoogleAccountsCallbackProxy(long tab, IGoogleAccountsCallbackClient client) {
        assert client != null;
        mClient = client;
        mNativeGoogleAccountsCallbackProxy =
                GoogleAccountsCallbackProxyJni.get().createGoogleAccountsCallbackProxy(this, tab);
    }

    public void setClient(IGoogleAccountsCallbackClient client) {
        assert client != null;
        mClient = client;
    }

    public void destroy() {
        GoogleAccountsCallbackProxyJni.get().deleteGoogleAccountsCallbackProxy(
                mNativeGoogleAccountsCallbackProxy);
        mNativeGoogleAccountsCallbackProxy = 0;
    }

    @CalledByNative
    private void onGoogleAccountsRequest(@GAIAServiceType int serviceType, String email,
            String continueUrl, boolean isSameTab) throws RemoteException {
        mClient.onGoogleAccountsRequest(
                implTypeToJavaType(serviceType), email, continueUrl, isSameTab);
    }

    @CalledByNative
    private String getGaiaId() throws RemoteException {
        return mClient.getGaiaId();
    }

    @GoogleAccountServiceType
    private static int implTypeToJavaType(@GAIAServiceType int type) {
        switch (type) {
            case GAIAServiceType.GAIA_SERVICE_TYPE_SIGNOUT:
                return GoogleAccountServiceType.SIGNOUT;
            case GAIAServiceType.GAIA_SERVICE_TYPE_ADDSESSION:
                return GoogleAccountServiceType.ADD_SESSION;
            // SIGNUP and INCOGNITO should not be possible currently, so pass through to DEFAULT.
            case GAIAServiceType.GAIA_SERVICE_TYPE_SIGNUP:
            case GAIAServiceType.GAIA_SERVICE_TYPE_INCOGNITO:
            case GAIAServiceType.GAIA_SERVICE_TYPE_DEFAULT:
                return GoogleAccountServiceType.DEFAULT;
        }
        assert false;
        return GoogleAccountServiceType.DEFAULT;
    }

    @NativeMethods
    interface Natives {
        long createGoogleAccountsCallbackProxy(GoogleAccountsCallbackProxy proxy, long tab);
        void deleteGoogleAccountsCallbackProxy(long proxy);
    }
}

File: weblayer/browser/google_accounts_callback_proxy.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "weblayer/browser/google_accounts_callback_proxy.h"

#include "base/android/jni_string.h"
#include "components/signin/core/browser/signin_header_helper.h"
#include "weblayer/browser/java/jni/GoogleAccountsCallbackProxy_jni.h"
#include "weblayer/public/tab.h"

using base::android::AttachCurrentThread;
using base::android::ConvertUTF8ToJavaString;

namespace weblayer {

GoogleAccountsCallbackProxy::GoogleAccountsCallbackProxy(JNIEnv* env,
                                                         jobject obj,
                                                         Tab* tab)
    : tab_(tab), java_impl_(env, obj) {
  tab_->SetGoogleAccountsDelegate(this);
}

GoogleAccountsCallbackProxy::~GoogleAccountsCallbackProxy() {
  tab_->SetGoogleAccountsDelegate(nullptr);
}

void GoogleAccountsCallbackProxy::OnGoogleAccountsRequest(
    const signin::ManageAccountsParams& params) {
  JNIEnv* env = AttachCurrentThread();
  Java_GoogleAccountsCallbackProxy_onGoogleAccountsRequest(
      env, java_impl_, params.service_type,
      ConvertUTF8ToJavaString(env, params.email),
      ConvertUTF8ToJavaString(env, params.continue_url), params.is_same_tab);
}

std::string GoogleAccountsCallbackProxy::GetGaiaId() {
  return base::android::ConvertJavaStringToUTF8(
      Java_GoogleAccountsCallbackProxy_getGaiaId(AttachCurrentThread(),
                                                 java_impl_));
}

static jlong JNI_GoogleAccountsCallbackProxy_CreateGoogleAccountsCallbackProxy(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& proxy,
    jlong tab) {
  return reinterpret_cast<jlong>(
      new GoogleAccountsCallbackProxy(env, proxy, reinterpret_cast<Tab*>(tab)));
}

static void JNI_GoogleAccountsCallbackProxy_DeleteGoogleAccountsCallbackProxy(
    JNIEnv* env,
    jlong proxy) {
  delete reinterpret_cast<GoogleAccountsCallbackProxy*>(proxy);
}

}  // namespace weblayer
